hc_dpki/src/keypair.rs

```diff
@@ -0,0 +1,551 @@
+use crate::{
+    bundle,
+    holochain_sodium::{aead, kx, random::random_secbuf, secbuf::SecBuf, sign},
+    util,
+};
+use holochain_core_types::{agent::KeyBuffer, error::HolochainError};
+use rustc_serialize::json;
+use std::str;
+
+pub struct Keypair {
+    pub pub_keys: String,
+    pub sign_priv: SecBuf,
+    pub enc_priv: SecBuf,
+}
+
+pub const SEEDSIZE: usize = 32 as usize;
+
+impl Keypair {
+    /// derive the pairs from a 32 byte seed buffer
+    ///  
+    /// @param {SecBuf} seed - the seed buffer
+    pub fn new_from_seed(seed: &mut SecBuf) -> Result<Self, HolochainError> {
+        let mut seed = seed;
+        let mut sign_public_key = SecBuf::with_insecure(sign::PUBLICKEYBYTES);
+        let mut sign_secret_key = SecBuf::with_secure(sign::SECRETKEYBYTES);
+        let mut enc_public_key = SecBuf::with_insecure(kx::PUBLICKEYBYTES);
+        let mut enc_secret_key = SecBuf::with_secure(kx::SECRETKEYBYTES);
+
+        sign::seed_keypair(&mut sign_public_key, &mut sign_secret_key, &mut seed)?;
+        kx::seed_keypair(&mut seed, &mut enc_public_key, &mut enc_secret_key)?;
+
+        Ok(Keypair {
+            pub_keys: util::encode_id(&mut sign_public_key, &mut enc_public_key),
+            sign_priv: sign_secret_key,
+            enc_priv: enc_secret_key,
+        })
+    }
+
+    /// get the keypair identifier string
+    ///
+    /// @return {string}
+    pub fn get_id(&mut self) -> String {
+        return self.pub_keys.clone();
+    }
+
+    /// generate an encrypted persistence bundle
+    ///
+    /// @param {SecBuf} passphrase - the encryption passphrase
+    ///
+    /// @param {string} hint - additional info / description for the bundle
+    pub fn get_bundle(
+        &mut self,
+        passphrase: &mut SecBuf,
+        hint: String,
+    ) -> Result<bundle::KeyBundle, HolochainError> {
+        let mut passphrase = passphrase;
+        let bundle_type: String = "hcKeypair".to_string();
+        let kk = KeyBuffer::with_corrected(&self.pub_keys)?;
+        let sk = kk.get_sig() as &[u8];
+        let ek = kk.get_enc() as &[u8];
+        let mut sk_buf = SecBuf::with_insecure(32);
+        let mut ek_buf = SecBuf::with_insecure(32);
+        util::convert_array_to_secbuf(&sk, &mut sk_buf);
+        util::convert_array_to_secbuf(&ek, &mut ek_buf);
+
+        // Merge all the secbuf together before encoding
+        let mut sign_pub = sk.to_vec();
+        let mut enc_pub = ek.to_vec();
+        let mut sign_priv = self.sign_priv.read_lock().to_vec();
+        let mut enc_priv = self.enc_priv.read_lock().to_vec();
+
+        sign_pub.append(&mut enc_pub);
+        sign_pub.append(&mut sign_priv);
+        sign_pub.append(&mut enc_priv);
```

Comment:
close! : ) except, these private keys have now been placed in unprotected memory.

Think we need to create a secure secbuf of the proper length, then copy in the contents of the various parts to the proper offsets. (also, I think we need that version byte at the beginning, incase we decide to change the bundle format in the future... though mabye that could be handled in the higher level with the `hint`?)

It may be worth creating a utility function that can copy `&[u8]`s into a secbuf at an offset, this could used in a lot of places around here, including the `from_parts` in secbuf itself

Reply:
Done